### DISCUSSION TITLE
indexing ZIM

### DISCUSSION THREAD

**Discussion by CRZeiss** [zim-desktop-wiki/zim-desktop-wiki#2229] [Q&A]
Hello,

I am new to ZIM and I am planning and working on bringing all my OneNote pages over to ZIM.
Therefore I created 7 notebooks (.zim) in my existing explorer structure and built around 150 zim-pages ("copied" from my OneNote-pages).

I am working on Windows 11.
The part of the filesystem in which ZIM is searching for notebooks and pages has around 3.000 subfolders and 240.000 files, most of them are .pdf, .msg (Outlook-messages), .jpg, .url and .lnk.

After calling the main notebook this morning ZIM began indexing the notebooks.
The indexing is running since this morning 5:00 and is still running now, 14:30 (since 9:30 hours), and I cannot see any progress anymore.

ZIM created a log-file zim.log, where I could watch many actions, but the last entry was at 8:40 this morning.
There is a folder named "...AppData\Roaming\zim\cache\zim\notebook-G\Cache" where I can see a file index.db (size 47.5 MB). Every 3-5 seconds a new file appears and disapears again (cannot read the filename). Size of the index.db didn't change since this morning.

I don't know what to do now, waiting till "the end" or breaking?
Can anybody tell me how to find out what ZIM is doing or what to do? 

Carl

**Comment by jaap-karssenberg** [comment]
Hi Carl,

For some background on expected indexing performance see: Performance large
notebook · zim-desktop-wiki/zim-desktop-wiki Wiki (github.com)
<https://github.com/zim-desktop-wiki/zim-desktop-wiki/wiki/Performance-large-notebook>

A structure with 240_000 files sounds like a huge data set. However you
mention only 150 pages created. Did you actually mean to index all this
data - e.g. are these all attachments to the 150 pages - or did you put zim
to work on a large folder structure ?

Technically speaking I would expect zim to only read the pages (".txt"
files that start with a specific header -- assuming you have a recent
version) so the amount of files should not matter as they are only listed,
which should go relatively fast.

Are there any other reasons why the system might get stuck when an
application tries to scan all files -- e.g. is it local file system, or
dynamically cloud synchronized ?

If you think zim is hanging, you should just close the process and try
again. Of course trying again means re-scanning a number of files. But it
should only scan folder listings and modification times to compare to the
indexed data. So if things work as intended that should not take hours.

Regards,

Jaap

**Comment by CRZeiss** [comment]
Hello Jaap,

Thank you for your help with my trouble.

The about 240.000 files have nothing to do with ZIM!
It is the "normal" content of my file-system, collections of images ,jpg and videos of about 20 years, collection of music-files mp3 of 100s of bands, 1000s of Excel-files, pdf-files, Word-files, email-messages .msg of about 30 years work as freelancer, and so on. Surlely some attached images to ZIM-pages, but these are .png.
My target is to integrate ZIM in my existing folder structure:
- for every client and procect there is a folder
- for every development project there is a folder  
- for every equipment, software, workflow there is a folder with links, infos and pdfs
- for every private person, project, insurance, idea there is a folder 

I _started_ some days ago to built the ZIM-files, actually I built about 150. My OneNote-notebook consists of about 2.000 to 2.500 pages. This is the target (there is much work to do!).

Ah, one speciality: I decided to use .txz files instead of .txt files for ZIM, so I have assigned the file-extension .txz to the ZIM application and can call every single ZIM-page directly with a .lnk-file. This information is included in every .zim file; default_file_extension=.txz

The filesystem is local, synchronized with Microsoft OneDrive, but this should not affect the speed of reading the local files.

> I would expect zim to only read the pages (".txt" files that start with a specific header)
That is what I expected. But in the zim.log I find every file type like msg, pdf, doc?, xls? lnk, jpg:
DEBUG: Index file: G:\Verwaltung\so_arc\CRZ\2017.1\Warnec-Eing.-Vertrag-15042-Fritz-.msg
DEBUG: Index file: G:\Verwaltung\so_arc\CRZ\2017.2\1&1 In_ER_171128 ER 1&1.pdf
DEBUG: Index file: G:\Verwaltung\so_arc\CRZ\2017.1\esd el_Lasts_Word.doc
...
How can I tell ZIM to analyze only the .txz-files? 

> If you think zim is hanging, you should just close
No, I think ZIM is still working, because in the cache-folder I see movement: every 1 to 5 seconds a new file is created and deleted again.

Some folders ZIM cannot work with:
AssertionError: Not a valid page name: #Handbuch CRUZ:10 Testamente-Verfügungen

While building a mostly new ZIM structure I assume that there will surely be some re-structuring (movements of ZIM-files) which needs re-indexing again. So I need a "solution" so that this re-indexing does not long for hours.

Carl

**Comment by CRZeiss** [comment]
Maybe this creating and deleting of file in the cache-folder depends on the file size of the analyzed (240.000) files.
If every single file is analyzed and needs 1-5 seconds, ZIM will work about 200 hours ...

So again the question of all questions: How can I tell ZIM to analyze only the .txz-files?

Carl

**Comment by jaap-karssenberg** [comment]
Maybe you should start running zim only on the folder of 150 pages +
attachments, and not try to scan your whole filesystem ?

--- Jaap

**Comment by CRZeiss** [comment]
> Maybe you should start running zim only on the folder of 150 pages
There is **no folder which contains only ZIM-files**.

This was the point: 
I wanted to **integrate** the ZIM-files in my **existing folder- and file-structure**, because all of my 300 client folders, my yearly document folders with documents, messages etc., all of the project folders and all of the development folders for 100 programs **already exist** and are filled with documents.
I did _not_ want to create 500 _new_ (mostly identical) folders at another place) and then have to create links from new ZIM-folders to existing client folders and back.

The idea is: 
ZIM should be an wiki **integrated in my my file- and information management**, not an independant detached wiki with no connection to the rest of my system. Otherwise I could have stayed with OneNote.

So again the main question:
is there a possibility or could you create a possibility that ZIM only scans the ZIM-files (.zim and .txz)?

By the way: the indexing is meanwhile finished, after about 13 hours. 
ZIM looks and works well.
I don't know when ZIM wants to re-indexing again...

Carl

**Comment by CRZeiss** [comment]
Task-Manager told me ZIM is working in the background now - using about 35% of CPU.
There is a new log-file which is much smaller than the one of the indexing.

I think it would be a good idea to combine each line in the log-file with the actual time.

**Comment by jaap-karssenberg** [comment]
I understand that you want to integrate with an existing structure. But you
might still want to do smaller bites. Zim is organized in "notebooks" so
you could choose to have separate notebooks per project or per activity.
Since you mentioned also eg music was in this structure you could consider
sub sets of your structure as separate notebooks.

Zim does two things while indexing: it reads the contents of the pages. All
other files it only takes file listing and timestamps. It will reindex
occasionally, eg when upgrading, because the index is essentially a throw
away cache.

In your case the amount of files is much larger than anything I've tested.
However, I would not expect many hours to just take file listing.

That is why I'm wondering if there is any reason the file system allows
down. Hard to debug your specific case without having a setup myself hands
on.

The best way to see what is happening during indexing is too run from
terminal with --index and --verbose options. But you will likely need an
msys2 environment to run this.

Regards,

Jaap

Op vr 28 okt. 2022 19:44 schreef CRZeiss ***@***.***>:

**Comment by CRZeiss** [comment]
10. Oktober 2022

 

Hi Jaap,

 

I am constantly working with Zim since about 3 weeks now.

More or less I am organizing my whole work and also private life with the wikis (Zim, former OneNote).

 

Notebooks

I am not working with the main wiki (240T files) anymore because of the very long indexing time.

(The first indexing of my main wiki needed about 13 hours. This happened automatically for 2 times 3 weeks ago)

 

You suggested to part my "main wiki" in several smaller wikis. 

I created 7 "sub-wikis" with which I am working since 3 weeks.

 

Indexing

The Zim-indexing works over all files and file types (see attached zim.log of one of the sub wikis).

After this first indexing (3 weeks ago with the big “main wiki”) I observed the behaviour of Zim:

- Zim is indexing every used notebook in the background from start of Zim and also after closing the application, needs about 30-40% of the CPU for hours, needs fan (noisy) and energy.

- In this moment (2 sub wikis are active) there are 3 cache files (= notebooks) in progress.

In total sum the number of files to re-index with the sub wikis is the same as using the main wiki (=240T files).

 

Searching for contents

Searching for content also seems to search the content of all the indexed (240T) files.

This is completely unnecessary, because I only want to search for content in all the files (=pages) belonging to the wiki.

One of the sub-wikis actually has only 4 pages but needs hours to search for a word because of about 30T indexed files. Couldn't wait till the end.

Another of the sub-wikis has 20 pages and 1.500 indexed files, search for a single word worked 40 minutes with no result, I had to break before end.

Also the Backlink-Search does not work - here also should be searched only the Zim-pages.

Result: searching for content is not possible with my file structure because of eternal response time.

So also searching should be restricted to the Zim-files (.txz).

(Workaround: I can seek use the explorer to seek for the contents of all .txz-files)

 

Debug log

Would it be possible to add the actual time of every log entry in the logfile? Would be useful to find out time required for single actions.

It seems that not all of the background activities are logged:

- In this moment there are 3 cache files in progress, but no logfile is logging these activities.

- during the first indexing (3 weeks ago) the logging ended after about 3 hours, but the cache file was in progress for another 10 hours.

So most of the progress time is not logged. Could this please be changed?

 

Problems and possible solution

My first problem - time needed for indexing of about 240.000 files - is not really solved by using the sub wikis.

Second problem - searching for contents in the pages would be necessary but cannot be used in my environment - no solution.

Maybe an additional debugging (see above) could clear the reasons for these problems.

 

I assume both problems could be solved by creating an additional option in the notebook.zim for indexing and searching file types (extensions).

 

I am aware that Zim was made to handle its own folder structure which includes more or less only Zim-files and attachements.

Can we talk about a payment/fee for these needed additions?

 

Carl

 

 

Von: Jaap Karssenberg ***@***.***> 
Gesendet: Samstag, 29. Oktober 2022 10:50
An: zim-desktop-wiki/zim-desktop-wiki ***@***.***>
Cc: CRZeiss ***@***.***>; Author ***@***.***>
Betreff: Re: [zim-desktop-wiki/zim-desktop-wiki] indexing ZIM (Issue #2213)

 

I understand that you want to integrate with an existing structure. But you
might still want to do smaller bites. Zim is organized in "notebooks" so
you could choose to have separate notebooks per project or per activity.
Since you mentioned also eg music was in this structure you could consider
sub sets of your structure as separate notebooks.

Zim does two things while indexing: it reads the contents of the pages. All
other files it only takes file listing and timestamps. It will reindex
occasionally, eg when upgrading, because the index is essentially a throw
away cache.

In your case the amount of files is much larger than anything I've tested.
However, I would not expect many hours to just take file listing.

That is why I'm wondering if there is any reason the file system allows
down. Hard to debug your specific case without having a setup myself hands
on.

The best way to see what is happening during indexing is too run from
terminal with --index and --verbose options. But you will likely need an
msys2 environment to run this.

Regards,

Jaap

Op vr 28 okt. 2022 19:44 schreef CRZeiss ***@***.*** <mailto:***@***.***> >:

**Comment by supertoothy** [comment]
Coming in late to this discussion. Don't have solutions, just offering my experience.

My use case is similar, and I also use Zim's folder structure as a way to organise all my files. I use Linux Mint. In my case though, I started my file structure organically with Zim many years ago and eliminated all other user folders outside of it (other than system defined ones like Templates, Downloads etc.)

I also have a lot of files to search through, so I don't use Zim's search for this at all. I use Recoll, which has indexed all 3 of my Notebooks with a total size of about 60GB at present. When I know where to find something, I use Zim's search by section. 

However, because of the way I have sectioned my notes, the parts I need to use Zim's searches on are mostly current work. Searching by section works well here. The older files are in an archival section, which is what Recoll combs through. I only use this rarely. Even then, I generally use ripgrep on emacs. Recoll is mostly for those occasions for when I know a concept i'm looking for is buried in a pile of PDFs and folders and I need to retrieve it.

Maybe someone could build on a recoll plugin for ZIm? Outsource complex searches with large datasets to a specialised indexing software and use the results? People with a technical background can comment on the feasability of this.

In general, my observations from long term use are:

-  For large repositories of notes, Zim's features by themself are not enough. I use Recoll for search and emacs as source editor. Both these helper tools deal with more complex use cases for searching and writing.
- Zim seems to work best for Notebooks with about a max of 5k notes in them. At this size it is still snappy and responsive with jumps or searches
- Good page naming conventions, annotating the page if it has attachments etc make retrieval much easier.

**Reply by introt** [reply]
Zim's search could be a lot faster even without an index (#64), but for "complex searches with large datasets" I definitely think a plugin wrapping a dedicated tool is the right answer

**Reply by CRZeiss** [reply]
Hello Jaap,

You remember: I am the one who wanted to integrate Zim in a file system with about 3.000 subfolders and 240.000 files..,
I started this discussion, because Zim was indexing all of these files and needed hours to do so.

I am now working for 4 months with Zim. 
My experiences in the meantime:

I am so far quite satisfied working with Zim.
I am working with 1 central notebook (consisting of 10 sub-notebooks), about 700 Zim-files (txt), included in my file system with about 3.000 subfolders and 250k files (rising).
Basic idea was and is: folder system structure = Zim Wiki structure.

For searching contents of Zim in this context the integrated search is not suitable. It tries to search not only the Zim txt-files but also the the contents of all 250k files, and this cannot work satisfying - every single search needs hours and doesn't come to an end.
So I am working with Instant Search. This Plugin works fast enough to satisfy my requirment, because it uses the Zim page index and searches only the content of the 700 Zim-txt-files).

The now remaining problem is that Zim is continously indexing the whole file system in the background because all 250k files are seen as "attachements" to Zim.
Although Zim is quitted, it is running the WHOLE DAY LONG in the background constantly using 30-40% of computers memory.
Active applications running slower, fan is running hot, battery is running empty - for indexing files which in my idea have nothing to do with Zim.
I do not really know what Zim is working on in the background. The Zim.log does not include most of these aktivities.
I can only see that hours after the last entry in zim.log every second a file appears and disapears in the cache folder and the date/time of index.db is changing.

I suggested to expand the logging so we can find out what Zim is working.
I also suggested to create an option which file types should be indexed (and searched) by Zim, for example only all txt-files = content of the Wiki without any attachements.
Unfortunaly on these tow suggestions I got no answer...

Maybe I can bring life in this discussion again.
I see there are some more users here with similar thoughts...

Carl
